pkg/controllers/telemetry/controller.go

```diff
@@ -0,0 +1,266 @@
+/*
+Copyright 2024 The CloudPilot AI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package telemetry
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	metricsapi "k8s.io/metrics/pkg/apis/metrics"
+	metricsV1beta1api "k8s.io/metrics/pkg/apis/metrics/v1beta1"
+	metricsclientset "k8s.io/metrics/pkg/client/clientset/versioned"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/cloudpilot-ai/karpenter-provider-alibabacloud/pkg/cloudprovider"
+	"github.com/cloudpilot-ai/karpenter-provider-alibabacloud/pkg/controllers/interruption"
+	"github.com/cloudpilot-ai/karpenter-provider-alibabacloud/pkg/operator/options"
+)
+
+const (
+	defaultTelemetryEndpoint = "https://api.cloudpilot.ai/api/v1/telemetry"
+	resourcesInfoPath        = "/clusters/%s/resourcesInfo"
+	interruptionEventPath    = "/clusters/%s/interruptionEvent"
+
+	karpenterInitializedKey = "karpenter.sh/initialized"
+)
+
+type Controller struct {
+	kubeClient   client.Client
+	metricClient metricsclientset.Interface
+
+	once sync.Once
+}
+
+func NewController(kubeClient client.Client, metricClient metricsclientset.Interface) *Controller {
+	return &Controller{
+		kubeClient:   kubeClient,
+		metricClient: metricClient,
+		once:         sync.Once{},
+	}
+}
+
+func (c *Controller) Reconcile(ctx context.Context, node *corev1.Node) (reconcile.Result, error) {
+	c.once.Do(func() {
+		PushTelemetryClusterResourceInfo(ctx, options.FromContext(ctx).ClusterID, c.kubeClient, c.metricClient)
+	})
+
+	_, interrupted := lo.Find(node.Status.Conditions, func(condition corev1.NodeCondition) bool {
+		return condition.Type == interruption.ConditionTypeInstanceExpired && condition.Status == corev1.ConditionTrue
+	})
+
+	if interrupted {
+		PushInterruptionEvent(ctx, options.FromContext(ctx).ClusterID, node)
```

Comment:
Encrypt the cluster ID to avoid uploading sensitive data:
```
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
)

func main() {
	input := "aaa"

	// Compute SHA-256 hash
	hash := sha256.New()
	hash.Write([]byte(input))
	hashBytes := hash.Sum(nil)

	// Convert hash bytes to a hexadecimal string
	hashString := hex.EncodeToString(hashBytes)

	fmt.Println("SHA-256 Hash:", hashString)
}
```